### PR TITLE
Emacs syntax highlighting and file-type detection

### PR DESCRIPTION
**Pull request by esben-semmle**
Emacs support similar to the [Vim support](<https://github.com/Semmle/ql/pull/871>). Relevant for [Make syntax highlightning configs public](<https://jira.semmle.com/browse/QL-809>).

I emphasize that this is a simple Emacs mode, and that we have much more advanced, but untable, mode internally.

Ping @rneatherway as another Emacs person.

### CODE REVIEW

File: misc/emacs/ql-mode-base.el [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+;;
+;; Provides syntax highlightning, comment support, and a mode-specific
+;; keymap.
+
+;;; Code:
+
+(defconst ql--at-type-regex "\\_<@\\w+\\>")
+(defconst ql--predicate-regex "\\(\\_<\\w+\\(\\+\\|\\*\\)?\\_>\\)\\s-*(")
+(defconst ql--primitive-type-regex (regexp-opt '("int" "string" "float" "boolean" "date") 'symbols))
+(defconst ql--annotation-regex (regexp-opt '("abstract" "cached" "external" "final" "transient" "library" "private" "deprecated" "override" "query" "language" "bindingset") 'words))
```
Comment (rneatherway) [inline misc/emacs/ql-mode-base.el]: It probably makes sense to keep this consistent with the [vim list](https://github.com/Semmle/ql/pull/871/files#diff-feccfb634405a05309579436eefc55e0R2). This list additionally has `transient`, while vim has:

```
monotonicAggregates
noinline
nomagic
noopt
pragma
```

Comment (esben-semmle) [reply misc/emacs/ql-mode-base.el]: Fair, I have changed it such that we now highlight all arguments to the parameterized annotations `pragma`, `language` and `bindingset`. In that way, we do not have to update the highlightning rules when new pragmas or languages are introduced.

File: misc/emacs/ql-mode-base.el [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+;; Provides syntax highlightning, comment support, and a mode-specific
+;; keymap.
+
+;;; Code:
+
+(defconst ql--at-type-regex "\\_<@\\w+\\>")
+(defconst ql--predicate-regex "\\(\\_<\\w+\\(\\+\\|\\*\\)?\\_>\\)\\s-*(")
+(defconst ql--primitive-type-regex (regexp-opt '("int" "string" "float" "boolean" "date") 'symbols))
+(defconst ql--annotation-regex (regexp-opt '("abstract" "cached" "external" "final" "transient" "library" "private" "deprecated" "override" "query") 'words))
+(defconst ql--parameterized-annotation-regex "\\<\\(pragma\\|language\\|bindingset\\) *\\[ *[a-z, ]*\\]")
```
Comment (rneatherway) [inline misc/emacs/ql-mode-base.el]: I think that `a-z` should also allow uppercase for `monotonicAggregate` to work.

It's slightly unfortunate that `bindingset[this]` and `bindingset[result]` are not highlighted correctly, although I'm not sure how important you feel that is.

Comment (esben-semmle) [reply misc/emacs/ql-mode-base.el]: Ok. I have simplified it. It is now consistent  with Vim.

Example display when loaded in `emacs -q`: 
image

Comment (rneatherway) [reply misc/emacs/ql-mode-base.el]: Great, this looks good to me locally.

**Pull request merged**